### PR TITLE
BQ Storage Write API Support 

### PR DESCRIPTION
**Pull request by Goutam1511**
<h4> Summary (Short summary of what is being done) : </h4> This change adds the BQ Storage Write API Support from [DataflowTemplates Repository](https://github.com/GoogleCloudPlatform/DataflowTemplates/pull/834/files#diff-66ca2bbaddc8c2e3ae8abfb6a3ffbf5dbcbfbd6e2fb7ee4da102062159550cf7R688)
<br>
<h4> Description (Describe in detail the fix made) : </h4> Adding a deterministic coder to encode and decode TableRow class to String to enable Storage Write API
<br>
<h4> Bug ID (if any) : </h4>
<br>
<h4> Public Documentation (if any) : </h4>
<br>
<h4> TESTED (Test Cases with scenario and description - must have 1 positive and 1 negative scenario) : </h4>

- Enabled Storage Write API with exactly-once semantics and at-least once semantics
- Enabled Storage Write API with triggeringFrequency and numberOfStreams > 1
- Checked Streaming Write is working even with changes
- Checked if pipeline throws exception with Storage Write API number of streams > 1 but no triggeringFrequency provided

TO TEST : Existing issue of row duplication

### CODE REVIEW

File: .github/workflows/dlp-pipelines.yml [hunk 1]
```diff
@@ -254,6 +254,81 @@ jobs:
           done
           echo "# records in input CSV file are: $rc_orig."
           echo "Verified number of rows in ${{env.INPUT_FILE_NAME}}_pub_sub: $row_count."
+
```
[the remark this answers is not in the document]
Comment (chitara-01) [reply .github/workflows/dlp-pipelines.yml:257]: Also, were these changes tested on cases where we were getting duplicates to ensure it works there? You can refer to the [experiments](https://docs.google.com/spreadsheets/d/1xxwWLYGsPNMfpZ6CquaJ7_DVpJa6lIJ5Xhm1G8BS0AQ/edit?resourcekey=0-ExzIpDApP_RwZmIR0dzX-A#gid=0) done previously for the same.

If already done, might be worth mentioning under TESTED.

File: .github/workflows/dlp-pipelines.yml [hunk 2]
```diff
@@ -254,6 +254,81 @@ jobs:
...
+          --useStorageWriteApi --storageWriteApiTriggeringFrequencySec=2 --numStorageWriteApiStreams=2"
+
+      - name: Verify BQ table for streaming write
+        env:
+          DATASET_ID: ${{ needs.generate-uuid.outputs.output5 }}
+        run: |
+          not_verified=true
+          table_count=0
+          while $not_verified; do
+            table_count=$(($(bq query --use_legacy_sql=false --format csv 'SELECT * FROM `${{env.PROJECT_ID}}.${{env.DATASET_ID}}`.__TABLES__ WHERE table_id LIKE "streaming_write%"' | wc -l ) -1))
```
Comment (chitara-01) [inline .github/workflows/dlp-pipelines.yml:301]: Can we mention the file/table name 'streaming_write' as a variable at the top instead of hard-coding it?

File: .github/workflows/dlp-pipelines.yml [hunk 3]
```diff
@@ -362,6 +437,7 @@ jobs:
       - generate-uuid
       - inspection
       - de-identification
+      - de-identification-streaming-write
```
Comment (chitara-01) [inline .github/workflows/dlp-pipelines.yml:440]: Is table created in 'de-identification-streaming-write' being cleaned-up here?

File: src/main/java/com/google/swarm/tokenization/common/Util.java [hunk 4]
```diff
@@ -387,13 +388,39 @@ public static TableRow createBqRow(Table.Row tokenizedValue, String[] headers) {
...
   }
 
+  public static void validateBQStorageApiOptionsStreaming(BigQueryOptions options) {
+    if (options.getUseStorageWriteApi()
+        && !options.getUseStorageWriteApiAtLeastOnce()
+        && (options.getNumStorageWriteApiStreams() < 1
+            || options.getStorageWriteApiTriggeringFrequencySec() == null)) {
+      // the number of streams and triggering frequency are only required for exactly-once semantics
+      throw new IllegalArgumentException(
+          "When streaming with STORAGE_WRITE_API, the number of streams"
```
Comment (chitara-01) [inline src/main/java/com/google/swarm/tokenization/common/Util.java:406]: Should it be explicitly mentioned in the exception that this is when using exactly once semantics?

File: src/main/java/com/google/swarm/tokenization/common/Util.java [hunk 5]
```diff
@@ -387,13 +388,39 @@ public static TableRow createBqRow(Table.Row tokenizedValue, String[] headers) {
...
+    validateStorageWriteApiAtLeastOnce(options);
+  }
+
+  private static void validateStorageWriteApiAtLeastOnce(BigQueryOptions options) {
+    if (options.getUseStorageWriteApiAtLeastOnce() && !options.getUseStorageWriteApi()) {
+      // Technically this is a no-op, since useStorageWriteApiAtLeastOnce is only checked by
+      // BigQueryIO when useStorageWriteApi is true, but it might be confusing to a user why
+      // useStorageWriteApiAtLeastOnce doesn't take effect.
+      throw new IllegalArgumentException(
+          "When at-least-once semantics (useStorageWriteApiAtLeastOnce) are enabled Storage Write"
```
Comment (chitara-01) [inline src/main/java/com/google/swarm/tokenization/common/Util.java:419]: Please correct the sentence to "When at-least-once semantics (useStorageWriteApiAtLeastOnce) is enabled, Storage Write" + " API (useStorageWriteApi) must also be enabled."

File: src/main/java/com/google/swarm/tokenization/DLPTextToBigQueryStreamingV2.java [hunk 6]
```diff
@@ -81,11 +82,13 @@ public static void main(String[] args) {
 
     DLPTextToBigQueryStreamingV2PipelineOptions options =
         PipelineOptionsFactory.fromArgs(args).as(DLPTextToBigQueryStreamingV2PipelineOptions.class);
+    Util.validateBQStorageApiOptionsStreaming(options);
```
Comment (chitara-01) [inline src/main/java/com/google/swarm/tokenization/DLPTextToBigQueryStreamingV2.java:85]: Were these changes tested for both exactly once and at least once semantics?

**Review by chitara-01 — CHANGES_REQUESTED** [review]

File: .github/workflows/dlp-pipelines.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (Goutam1511) [reply .github/workflows/dlp-pipelines.yml:257]: @chitara-01 Average time taken by entire workflow has increased because one of the job among inspection, de-identification and de-identification-streaming-write is waiting for the runner to finish others. We need to create one more self-hosted runner to make it efficient

[the remark this answers is not in the document]
Comment (Goutam1511) [reply .github/workflows/dlp-pipelines.yml:257]: > Also, were these changes tested on cases where we were getting duplicates to ensure it works there? You can refer to the [experiments](https://docs.google.com/spreadsheets/d/1xxwWLYGsPNMfpZ6CquaJ7_DVpJa6lIJ5Xhm1G8BS0AQ/edit?resourcekey=0-ExzIpDApP_RwZmIR0dzX-A#gid=0) done previously for the same.
> 
> If already done, might be worth mentioning under TESTED.

Will test this when comparing throughput and reliability improvements. For easy tracking, noting it down under PR description

File: src/main/java/com/google/swarm/tokenization/DLPTextToBigQueryStreamingV2.java (hunk 6 above)
Comment (Goutam1511) [reply src/main/java/com/google/swarm/tokenization/DLPTextToBigQueryStreamingV2.java:85]: Yes both of these cases have been tested. Thanks will mention the same in PR Description as well

File: .github/workflows/dlp-pipelines.yml (hunk 3 above)
Comment (Goutam1511) [reply .github/workflows/dlp-pipelines.yml:440]: Yes because ultimately the whole dataset is being removed

File: .github/workflows/dlp-pipelines.yml [hunk 7]
```diff
@@ -254,6 +254,81 @@ jobs:
...
+          java-version: 11
+
+      - name: Setup Gradle
+        uses: gradle/gradle-build-action@v2
+
+      - name: Run DLP Pipeline for streaming write
+        env:
+          DATASET_ID: ${{ needs.generate-uuid.outputs.output5 }}
+        run: |
+          gradle run -DmainClass=com.google.swarm.tokenization.DLPTextToBigQueryStreamingV2 -Pargs=" \
```
[the remark this answers is not in the document]
Comment (Goutam1511) [reply .github/workflows/dlp-pipelines.yml:284]: Done

File: src/main/java/com/google/swarm/tokenization/common/Util.java (hunk 4 above)
Comment (Goutam1511) [reply src/main/java/com/google/swarm/tokenization/common/Util.java]: Done

File: src/main/java/com/google/swarm/tokenization/common/Util.java (hunk 5 above)
Comment (Goutam1511) [reply src/main/java/com/google/swarm/tokenization/common/Util.java]: Done

**Review by chitara-01 — CHANGES_REQUESTED** [review]

**Review by chitara-01 — APPROVED** [review]
LGTM
